**<h1><center>Laboratorio 11: LLM y Agentes Autónomos 🤖</center></h1>**

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### **Cuerpo Docente:**

- Profesor: Ignacio Meza, Sebastian Tinoco
- Auxiliar: Catherine Benavides, Consuelo Rojas
- Ayudante: Eduardo Moya, Nicolás Ojeda

### Equipo: **SUPER IMPORTANTE - notebooks sin nombre no serán revisados**

- Nombre de alumno 1: Sergio Rehbein
- Nombre de alumno 2: Matías Cornejo

**SUPER IMPORTANTE** - notebooks sin nombre no serán revisados

\### **Link de repositorio de GitHub Matias:** https://github.com/s-kill/MDS7202

\### **Link de repositorio de GitHub Sergio:** https://github.com/sergiorehbein/MDS7201---Proyecto-de-Ciencia-de-Datos

### **Indice**

1. [Temas a tratar](#Temas-a-tratar:)
3. [Descripcción del laboratorio](#Descripción-del-laboratorio.)
4. [Desarrollo](#Desarrollo)

## **Temas a tratar**

- Implementación de modelos de LLM y Reinforcement Learning.
- Utilización e implementación de agentes.

## **Reglas:**

- **Grupos de 2 personas**
- Fecha de entrega: 7 días desde la publicación, 3 días de atraso con 1 punto de descuento c/u.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- Prohibidas las copias. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
Pueden usar cualquer material del curso que estimen conveniente.

### **Objetivos principales del laboratorio**

- Generar un modelo LLM generativo interactivo.
- Entrenar un modelo de Reinforce Learning.

El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

## **1. Large Language Models (4.0 puntos)**

<p align="center">
  <img src="https://trestristescriticos.com/wp-content/uploads/2021/07/telefono-gratuito-cinesur.jpg" width="350">
</p>

Joaquín no es un aficionado del cine, pero a principios de año, se propuso ver más peliculas para poder tener más temas de conversación con sus amigos y familia. Sin embargo, ya es junio y Joaquín no ha visto ninguna pelicula nueva o relevante de las que tenía en su lista y su reunión familiar bi-anual se acerca y necesita la mayor información que pueda recopilar de dichas peliculas sin tener que verlas.

Para esto, usted con su compañerx, tendrá que crear una aplicación utilizando LangChain.


---



**Instalación de librerías**

Para la creación de la aplicación, se utilizara un modelo de lenguaje (LLM) ofrecido gratuitamente por Google.

Para ello, se utilizará la API de Gemini, por lo que si no tienen acceso, se pueden crear una cuenta en el siguiente [enlace a Google AI](https://ai.google.dev/). Ahí, ir a la pestaña superior y seleccione la opción que dice ``Gemini API``.

<img src='https://gitlab.com/imezadelajara/datos_clase_7_mds7202/-/raw/main/misc_images/Screenshot_2024-06-13_at_12.42.32_PM.png' width='450' />

Luego, seleccione el botón que dice ``Get API key in Google AI Studio`` y hacer click en ``Crear clave de API`` para generar la llave con la que se podrá consultar al modelo de lenguaje.

<img src='https://gitlab.com/imezadelajara/datos_clase_7_mds7202/-/raw/main/misc_images/Screenshot_2024-06-13_at_12.45.10_PM.png?ref_type=heads' width='450' />

**Importante:** Debido a las restricciones de esta API, lo ideal es utilizar la llave a la API de manera personal.


Para mayor información sobre **LangChain**, pueden revisar la documentación en el [presente enlace](https://python.langchain.com/v0.2/docs/tutorials/summarization/ ).

In [2]:
%%capture
!pip install langchain
!pip install langchain_google_genai
!pip install langchain-community
!pip install langchain-experimental
!pip install sentence-transformers
!pip install faiss-cpu

In [3]:
from langchain_google_genai import ChatGoogleGenerativeAI
import os

os.environ["GOOGLE_API_KEY"] = "AIzaSyDB95Ktc4MiqMcz8Z6xFx108_wpylgQTqo" #AIzaSyDB95Ktc4MiqMcz8Z6xFx108_wpylgQTqo
llm = ChatGoogleGenerativeAI(model="gemini-1.5-flash")

### **1.1 Carga y limpieza (0.5 puntos)**

Para iniciar su titanica tarea de enseñarle a Joaquín sobre las mejores peliculas del último tiempo, tiene que revisar los script de las siguientes 3 peliculas:
* Dune 2
* Under Paris
* Joker

Debe encontrar un patrón y obtener solamente el guión de las películas. Para ello se recomienda utilizar métodos de búsqueda y reemplazo que tienen los ``string`` en Python. Adicionalmente, puede usar filtros de expresiones regulares.

Posterior a la limpieza de los guiones, debe considerar que el patrón se repite y es generalizable.


In [4]:
# Scripts de peliculas
dune2_script="https://scrapsfromtheloft.com/movies/dune-part-two-2024-transcript/"
underparis_script="https://scrapsfromtheloft.com/movies/under-paris-2024-transcript/"
joker_script="https://scrapsfromtheloft.com/movies/joker-2019-transcript/"

In [5]:
from langchain.document_loaders import WebBaseLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
import re

def load_website_data(url):

  loader = WebBaseLoader(url)
  website_data = loader.load()

  return website_data[0].page_content

def remove_text_before_marker(text):
    match = re.search("\t\t\n\n\n\n \n\n\n\n", text)
    body = text[match.end():]

    if "* * *" in body:
        match = re.search("\* \* \*", body)

        text_out = body[match.end():]
    else:
        text_out = body

    match = re.search('\n\n\n\n\n\n\n\n\n\n\n\n\n', text_out)
    text_out = text_out[:match.start()]
    return text_out

In [6]:
dune2 = load_website_data(dune2_script)
underparis = load_website_data(underparis_script)
joker = load_website_data(joker_script)

In [7]:
dune2_text = remove_text_before_marker(dune2)
underparis_text = remove_text_before_marker(underparis)
joker_text = remove_text_before_marker(joker)

### **1.2 Aplicación (3.5 puntos)**

Luego de limpiar los guiones, es posible generar la aplicaicón deseada con el LLM. Esta aplicación tiene que ser capaz de realizar las siguientes tareas.

1. Utilizando una plantilla sobre el nombre del archivo o la URL, identifique el supuesto nombre de la película.

2. Genere un resumen en español de la película y una nota evaluativa sobre la misma. El resumen debe tener entre 3 a 5 párrafos. Además, obtener una evaluación de la película con una calificación del 1 al 10, utilizando una LLM y el contexto entregado

#### **1.2.1 Título de la película (0.5 puntos)**

Para obtener el título, utilicé la siguiente plantilla:
```
 template = """
  What is the movie that appears in the description of this file or url?
  You only give me the movie name, nothing more.
  document/url: {script_path_url}
  """
```



In [8]:
def get_movie_title(script_path_url):
  template = """
  What is the movie that appears in the description of this file or url?
  You only give me the movie name, nothing more.
  document/url: {script_path_url}
  """
  prompt = PromptTemplate.from_template(template)
  prompt_val = prompt.invoke({"script_path_url": script_path_url})
  result = llm.invoke(prompt_val)

  return result.content.strip()

#### **1.2.2 Resumen (1.0 puntos)**

Como se vió en clases, las LLM no pueden manejar cadenas de texto muy largas, esto es debido a que, dependiendo de su naturaleza, solo manejan ventanas de contexto que estan asociadas a caracteristicas de la red y del entrenamiento utilizado.

Por ello, es altamente importante que si se desea hacer un resumen del texto, este se haga realizando un tipo de map/reduce sobre el texto. De manera que en cada una de las iteraciones se vaya disminuyendo el tamaño del texto, pero hay que tener cuidado con que le modelo vaya guardando el contexto de escenas previas.

In [9]:
from langchain.chains import MapReduceDocumentsChain, ReduceDocumentsChain
from langchain.chains import StuffDocumentsChain, LLMChain
from langchain.chains.combine_documents.stuff import StuffDocumentsChain
from langchain_core.prompts import PromptTemplate
from langchain_text_splitters import RecursiveCharacterTextSplitter


In [10]:
#No cambiar función

def map_reduce_text(script, map_template, reduce_template):

  # Map
  """
  map_prompt, crear el prompt desde el template
  map_chain, crear la cadena desde el prompt
  """
  map_prompt = PromptTemplate.from_template(map_template)
  map_chain = LLMChain(llm=llm, prompt=map_prompt)

  # Reduce
  """
  reduce_prompt, crear el prompt desde el template
  reduce_chain, crear la cadena desde el prompt
  """
  reduce_prompt = PromptTemplate.from_template(reduce_template)
  reduce_chain = LLMChain(llm=llm, prompt=reduce_prompt)

  # Combine
  """
  Combinar y reducir los documentos, utilizar StuffDocumentsChain
  y ReduceDocuentsChain con un máximo de 4000 tokens
  """
  combine_documents_chain = StuffDocumentsChain(
      llm_chain=reduce_chain,
      document_variable_name="document",
  )


  reduce_documents_chain = ReduceDocumentsChain(
      combine_documents_chain = combine_documents_chain,
      collapse_documents_chain = combine_documents_chain,
      token_max = 4000,
  )

  # Map/Reduce
  """
  Uilizar MapReduceDocumentsChain
  """

  map_reduce_chain = MapReduceDocumentsChain(
    llm_chain=map_chain,
    reduce_documents_chain=reduce_documents_chain,
    document_variable_name="document",
    return_intermediate_steps=False,
)

  # Text splitter
  """
  Usar RecursiveCharacterTextSplitter
  """
  text_splitter = RecursiveCharacterTextSplitter(
      chunk_size = 15000,
      chunk_overlap = 0
  )

  split_script = text_splitter.split_text(script)


  # resultado
  split_script = text_splitter.create_documents(split_script)

  #result = map_reduce_chain.run()
  #return result

  result = map_reduce_chain.invoke(split_script)

  return result["output_text"]

In [11]:
# crear templates

map_template_summary = """
Resume la siguiente parte del guion. Mantén los detalles importantes y el contexto:
{document}
"""

reduce_template_summary = """
Combina los siguientes resúmenes en un solo resumen cohesivo. Asegúrate de mantener los detalles importantes y el contexto:
{document}
"""

answer_summary = """
Genere un resumen en español del documento y una nota evaluativa sobre el mismo. El resumen debe tener entre 3 a 5 párrafos. Además, obtener una evaluación de la película con una calificación del 1 al 10.
Este resumen debe ser creado SOLO a partir de esta versión resumida del guion proporcionada a continuación. No debe incluir información externa o conocimiento previo sobre la película. Aquí está el guion resumido:
{document}
"""

In [ ]:
# imprimir resumenes de películas.
dune2_reduce_summary = map_reduce_text(dune2_text, map_template_summary, reduce_template_summary)
underparis_reduce_summary = map_reduce_text(underparis_text, map_template_summary, reduce_template_summary)
joker_reduce_summary = map_reduce_text(joker_text, map_template_summary, reduce_template_summary)

prompt = PromptTemplate.from_template(answer_summary)
dune2_resumen = llm.invoke(prompt.invoke({"document": dune2_reduce_summary}))
print(dune2_resumen.content)

underparis_resumen = llm.invoke(prompt.invoke({"document": underparis_reduce_summary}))
print(underparis_resumen.content)

joker_resumen = llm.invoke(prompt.invoke({"document": joker_reduce_summary}))
print(joker_resumen.content)

## Resumen de Dune

"Dune" nos introduce en un universo de intrigas políticas y luchas por el poder, centrado en el planeta desértico de Arrakis, hogar de la especia, una sustancia de gran valor que otorga longevidad y habilidades psíquicas. La historia sigue a Paul Atreides, un joven noble que se convierte en el salvador profetizado de los Fremen, un pueblo indígena de Arrakis. 

Tras la destrucción de la Casa Atreides a manos de los Harkonnen, sus enemigos, Paul y su madre, Jessica, se refugian entre los Fremen. Aunque inicialmente recelosos, los Fremen reconocen a Paul como el Lisan al-Gaib, un salvador profetizado, y a Jessica como la nueva Madre Superiora. Paul se integra a la cultura Fremen, aprendiendo sus costumbres y habilidades, ganándose el respeto de su líder, Stilgar. Se convierte en un Fedaykin, un guerrero Fremen, y recibe el nombre de Muad'Dib, "El que señala el camino".

Mientras tanto, los Harkonnen, liderados por el Barón Harkonnen y sus hijos, buscan destruir a los 

Adicionalmente, Joaquín sabe que su primo favorito le gusta ``Dune: Part 2`` por lo que le gustaría tener mayor información al respecto, para ello realice las siguientes tareas:


3. Genere un gráfico que muestre los personajes de la película con más apariciones en la misma.
4. Genere una tabla en pandas con los 3 personajes que más aparecen, indicando el nombre del actor y su edad actual más uno (ojo edad + 1).
5. Cree una función que responda preguntas sobre la película basándose en la información del texto entregado (OJO: las preguntas y salidas deben ser en español). Luego, responda las siguientes preguntas:
* ¿Qué y quién es Lisan al-Gaib?
* ¿Qué personaje no cree en la profecía pero es parte de ella?
* ¿Cuál es el objetivo de Feyd-Rautha?
6. Utilizando el top 3 de personajes que más aparecen en la película, genere con el modelo LLM y utilizando el contexto del guion, las 6 estadísticas que demuestren las habilidades de los personajes: Intelligence, Strength, Charisma, Wisdom, Emotional Resilience, y Creativity.

#### **1.2.3 Personajes (0.5 puntos)**

En la siguiente sección, tiene que entregar un template de personajes y redicción

In [ ]:
map_template_characters = """
Identifica los personajes en la siguiente parte del guion y cuenta sus apariciones. Incluye el nombre del personaje y el número de apariciones en este segmento:
{document}
"""

reduce_template_characters = """
Combina las siguientes listas de personajes y sus apariciones en una sola lista cohesiva. Suma las apariciones de cada personaje:
{document}

El resultado debe tener el siguiente formato JSON:
[
  {{"personaje": "Nombre del personaje", "apariciones": Número de apariciones}},
  ...
]
"""

answer_character_list = map_reduce_text(
    dune2_text,
    map_template_characters,
    reduce_template_characters
)

In [ ]:
import pandas as pd
import plotly.express as px
from itertools import count
import ast
import re
import json

def plot_characters(answer_character_list):
  # Clear answer
  answer_character_list = re.sub(r'\s+', ' ', answer_character_list).strip()[8:-3]
  character_pd = pd.DataFrame(json.loads(answer_character_list))

  # distil the characters output
  """
  Recomendación, utilizar un diccionario para ordenar los personajes
  """
  # Ordenar personajes por aparición
  character_pd = character_pd.sort_values(by='apariciones', ascending=False)


  # Create dataframe
  """
  De diccionario a DataFrame
  """
  Top3_characters = character_pd.head(3)

  # Graficar datos
  fig = px.bar(character_pd, x='personaje', y='apariciones', title='Apariciones de Personajes en la Película')
  fig.show()

  #Retornar los personajes
  return character_pd['personaje']


In [ ]:
lista_personajes = plot_characters(answer_character_list)

#### **1.2.4 Actores principales (0.75 puntos)**

Importante saber que el script **no** maneja información de los actores, por ello, es importante que nuestra LLM tenga acceso a internet, de manera de poder realizar búsquedas que nos ayuden a completar la información consultada.

Para esto, utilizaremos agentes combinados con react para realzar la consulta y asegurarnos de que la respuesta es correcta.

In [ ]:
from langchain.agents import load_tools
from langchain.agents import AgentType, initialize_agent

In [ ]:
# Key para realizar una busqueda
os.environ["SERPER_API_KEY"] = 'd63e62662ef63eb9e44ab133d191f7a99a0024a3'

In [ ]:
tools = load_tools(["google-serper", "llm-math"], llm=llm)
agent = initialize_agent(tools, llm, agent="zero-shot-react-description", verbose=True)

/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:139: LangChainDeprecationWarning:

The function `initialize_agent` was deprecated in LangChain 0.1.0 and will be removed in 0.3.0. Use Use new agent constructor methods like create_react_agent, create_json_agent, create_structured_chat_agent, etc. instead.



In [ ]:
lista_personajes.head(3)

3                         Stilgar
1                            Paul
6    Harkonnen Soldiers (general)
Name: personaje, dtype: object

In [ ]:
import datetime
def get_actors_and_age(character):

  # Inicializar tools y agente.
  tools = load_tools(["google-serper", "llm-math"], llm=llm)
  agent = initialize_agent(tools, llm, agent="zero-shot-react-description", verbose=True)

  # Crear template de query
  query_template_nombre = """
  Who plays {character} in Dune 2?. Answer must be only the actor name.
  """

  query_template_edad = """
  "What is the actor {actor_name} birth year?"
  """

  # Crear prompt y usar agente para la búsqueda.
  prompt_nombre = PromptTemplate.from_template(query_template_nombre)
  prompt_edad = PromptTemplate.from_template(query_template_edad)

  # Retornar Nombre y Edad + 1
  nombre_actor = agent.invoke(prompt_nombre.invoke({"character": character}))['output']
  edad_actor = agent.invoke(prompt_edad.invoke({"actor_name": nombre_actor}))['output']

  try:
    birth_year = int(edad_actor.strip())
    current_year = datetime.datetime.now().year
    actor_age_p1 = current_year - birth_year + 1
  except ValueError:
    actor_age_p1 = "Edad no encontrada"

  return nombre_actor, actor_age_p1

In [ ]:
nombre, edad = get_actors_and_age('Stilgar')



> Entering new AgentExecutor chain...
Thought: I need to find out who plays Stilgar in Dune 2. 
Action: google_serper
Action Input: Who plays Stilgar in Dune 2?
Observation: Javier Bardem
Thought:Thought: I now know the answer.
Final Answer: Javier Bardem 


> Finished chain.


> Entering new AgentExecutor chain...
Thought: I need to find the birth year of Javier Bardem. 
Action: google_serper
Action Input: Javier Bardem birth year
Observation: 1969
Thought:Thought: I now know the final answer
Final Answer: 1969 


> Finished chain.


In [ ]:
print(nombre, edad)

Javier Bardem 56


Tras distintas iteraciones de Prompts, descubrimos que preguntas con respuestas simples evitaban que google serper entrara en loop, por lo tanto decidimos separar la pregunta del nombre y la edad del actor por separado. También descubrimos que dado que es más sencillo obtener la fecha de nacimiento que la edad calculada, preguntamos por el año de nacimiento e hicimos la resta (más uno) para obtener la edad deseada.

#### **1.2.5 Personajes Stats (0.5 puntos)**

Esta parte es similar al punto 2. La clave esta en crear un buen prompting que nos permita generar las estadísticas basandonos en una búsqueda por map/reduce.

Tras la búsqueda, la idea es tener una función de Python que nos permita generar el gráfico deseado y tener el resumen de los personajes.


In [102]:
import json

def map_reduce_text(script, character):
  # Map
  map_template = """
  You are analyzing a movie script. Extract and analyze the following attributes for the character {character}:
  Intelligence, Charisma, Strength, Wisdom, Emotional Resilience, and Creativity.
  Provide your analysis in a clear and detailed manner.
  """

  # crear prompt y cadena
  #map_template += template_complemt
  map_prompt = PromptTemplate.from_template(map_template)
  map_chain = LLMChain(llm=llm, prompt=map_prompt)

  # Reduce
  reduce_template = """
  Based on the provided analysis, rate the character {character} on a scale from 1 to 10 for the following attributes:
  Intelligence, Charisma, Strength, Wisdom, Emotional Resilience, and Creativity.
  Provide a summary for each attribute explaining the rating in spanish.
  Besides the summary add a json format text:
  [
    {{"Intelligence": Intelligence Rate, "Charisma": Charisma rate, ...}},
    ...
  ]

  Split the summary from the json text with the text ***
  """
  reduce_prompt = PromptTemplate.from_template(reduce_template)
  reduce_chain = LLMChain(llm=llm, prompt=reduce_prompt)

  # Reduce
  """
  Reducir y combinar los documentos con un máximo de 4000 tokens
  """
  combine_documents_chain = StuffDocumentsChain(
      llm_chain=reduce_chain,
      document_variable_name="character",
  )

  reduce_documents_chain = ReduceDocumentsChain(
      combine_documents_chain = combine_documents_chain,
      collapse_documents_chain = combine_documents_chain,
      token_max = 4000,
  )

  # Map/Reduce
  """
  Uilizar MapReduceDocumentsChain
  """

  map_reduce_chain = MapReduceDocumentsChain(
    llm_chain=map_chain,
    reduce_documents_chain=reduce_documents_chain,
    document_variable_name="character",
    return_intermediate_steps=False,
  )


  # Text splitter
  """
  Usar RecursiveCharacterTextSplitter
  """
  text_splitter = RecursiveCharacterTextSplitter(
      chunk_size = 15000,
      chunk_overlap = 0
  )

  split_script = text_splitter.split_text(script)
  split_script = text_splitter.create_documents(split_script)

  # resultado
  result = map_reduce_chain.invoke(split_script)
  return result["output_text"]


# Formato del perfil
def format_profile(answer_character_profile):
  """
  Crear un json con las caracteristicas y que retorne
  (final_profile, stats) del personaje
  """
  match = re.search("\*\*\*", answer_character_profile)
  json_text = answer_character_profile[match.end():].strip()
  json_text = re.sub(r'\s+', ' ', answer_character_list[match.end():].strip()).strip()
  json_text = re.sub(r'\`', ' ', json_text).strip()
  json_text = re.sub(r'json', ' ', json_text).strip()

  stats = json.loads(json_text)
  final_profile = answer_character_profile[:match.start()]
  return (final_profile, stats[0])


In [79]:
answer_character_list = map_reduce_text(
    dune2_text,
    'Paul Atreides'
)

In [103]:
final_profile, stats = format_profile(answer_character_list)

In [107]:
print(final_profile)

## Character Attribute Analysis: Paul Atreides

Here's a breakdown of Paul's attributes on a scale of 1 to 10, with explanations for each rating:

**Intelligence:** 9

Paul demonstrates a remarkable understanding of complex political machinations, strategic planning, and cultural nuances. He quickly grasps the importance of spice control, the power dynamics of the Great Houses, and the Fremen culture. His ability to anticipate threats, manipulate others, and devise effective strategies demonstrates a sharp and strategic mind. 

**Charisma:** 8

Paul possesses a natural charisma that attracts followers and inspires loyalty. He commands respect through his confident leadership, bold declarations, and unwavering resolve. He uses his charm and manipulation skills to influence others, gaining their trust and support.

**Strength:** 7

Paul's physical prowess is demonstrated through his fighting skills, evident in his duel with Feyd-Rautha. However, his strength also lies in his mental forti

In [104]:
# Función para gráficar stats. No Tocar.

import plotly.graph_objects as go
import numpy as np

def plot_stats(stats, character_name="Paul Atreides"):
    base_stats = [
        "Intelligence", "Charisma", "Strength",
        "Wisdom", "Emotional Resilience", "Creativity"
    ]
    for stat in base_stats:
        if stat not in stats:
            stats[stat] = 0

    labels = list(stats.keys())
    stats_values = list(stats.values())
    stats_values += stats_values[:1]
    labels += labels[:1]

    # Plotly figure
    fig = go.Figure()
    fig.add_trace(go.Scatterpolar(
        r=stats_values,
        theta=labels,
        fill='toself',
        name=character_name
    ))

    fig.update_layout(
        polar=dict(
            radialaxis=dict(
                visible=True,
                range=[0, max(stats_values)]
            )
        ),
        showlegend=False,
        title=character_name
    )

    return fig

fig = plot_stats(stats)

In [105]:
fig.show()

#### **Comentar (0.25 puntos)**

En primer lugar, se utiliza un prompt para resumir el script, se especifica un enfoque particular para resolver la pregunta en cuestión.

Para la solución del problema utilizamos un prompt que indicaba especificamente una serie de caracteres para identificar el resumen del json. Facilitando la extracción respectiva.

1. ¿Qué otras tareas se podría realizar? De dos ejemplos con la metodología asociada.

Dado que la respuesta del modelo es Resumen *** json. Se puede utilizar para cualquier problema que requiera un conjunto de datos estandarizados.

Por ejemplo:


*   Analisis de género de la pelicula: Se podría usar un prompt que analice los distintos géneros de la pelicula y devuelta un json con un puntaje de proporción de cada genero. (Una pelicula con genero romantica puede tener romance, pero no ser especificamente de romance)
*   Lineas de participación de personajes: Similar a unos de los problemas anteriores, se puede hacer un conteo de la participación, a nivel de lineas en el script, para cada personaje.


2. ¿Cual es la importancia de los prompt y como estos afectan al desempeño de los LLM?

Los prompts guián al modelo para genear respuestas específicas. La idea es tener un prompt claro para reducir ambuguedades y mejorar precisión.

3. ¿Alguna de sus respuestas fue una 'alucinación'? ¿Por qué sucede esto?

Cuando utilizamos el modelo para generar un resumen del script de Dune parte 2 el modelo utilizó la información que conocía previamente, ya que el resumen generado para Dune parte 2 correspondia a un resumen de Dune parte 1.

Según lo investigado (chatGPT), esto pudo ocurrir porque el modelo debe haber prioirizado información ya conocida sobre Dune parte 1 en lugar de enforcarse en el nuevo guión proporcionado.




## **2. Reinforcement Learning (2.0 puntos)**

En esta sección van a usar métodos de RL para resolver dos problemas interesantes: `Blackjack` y `LunarLander`.

In [ ]:
!pip install -qqq gymnasium stable_baselines3
!pip install -qqq swig
!pip install -qqq gymnasium[box2d]

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 953.9/953.9 kB 3.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 182.3/182.3 kB 9.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.3/21.3 MB 46.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 8.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 374.4/374.4 kB 2.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done


### **2.1 Blackjack (1.0 puntos)**

<p align="center">
  <img src="https://www.recreoviral.com/wp-content/uploads/2016/08/s3.amazonaws.com-Math.gif"
" width="400">
</p>

Joaquín es fanático del Blackjack, por lo que en esta subsección implementarán métodos de RL y así generar una estrategia para que pueda ~~ir al casino a  hacerse millonario~~ aprender a resolver problemas mediante RL.

Comencemos primero preparando el ambiente. El siguiente bloque de código transforma las observaciones del ambiente a `np.array`:


In [ ]:
import gymnasium as gym
from gymnasium.spaces import MultiDiscrete
import numpy as np

class FlattenObservation(gym.ObservationWrapper):
    def __init__(self, env):
        super(FlattenObservation, self).__init__(env)
        self.observation_space = MultiDiscrete(np.array([32, 11, 2]))

    def observation(self, observation):
        return np.array(observation).flatten()

# Create and wrap the environment
env = gym.make("Blackjack-v1")
env = FlattenObservation(env)

#### **2.1.1 Descripción de MDP (0.2 puntos)**

Entregue una breve descripción sobre el ambiente [Blackjack](https://gymnasium.farama.org/environments/toy_text/blackjack/) y su formulación en MDP, distinguiendo de forma clara y concisa los estados, acciones y recompensas.

El ambiente de Blackjack en Gymnasium modela el juego de cartas Blackjack usando un Proceso de Decisión de Markov (MDP).

•	Estados: Consisten en una tupla de tres elementos: la suma actual del jugador, el valor de la carta visible del crupier (1-10), y si el jugador tiene un as utilizable (0 o 1).

•	Acciones: Discretas, con dos posibles: pedir carta (hit) o plantarse (stick).

•	Recompensas: +1 por ganar, -1 por perder, 0 por empatar y +1.5 por ganar con un Blackjack natural.

#### **2.1.2 Generando un Baseline (0.2 puntos)**

* Simule un escenario en donde se escojan acciones aleatorias. Repita esta
simulación 5000 veces y reporte el promedio y desviación de las recompensas.
* ¿Cómo calificaría el performance de esta política?
* ¿Cómo podría interpretar las recompensas obtenidas?

In [ ]:
import gymnasium as gym
from gymnasium.spaces import MultiDiscrete
import numpy as np

# Simular 5000 episodios con acciones aleatorias
n_episodes = 5000
rewards = []

for _ in range(n_episodes):
    done = False
    observation, _ = env.reset()
    total_reward = 0
    while not done:
        action = env.action_space.sample()  # Acción aleatoria
        observation, reward, done, _, _ = env.step(action)
        total_reward += reward
    rewards.append(total_reward)

# Calcular promedio y desviación estándar
mean_reward = np.mean(rewards)
std_reward = np.std(rewards)

print(f"Promedio de recompensas: {mean_reward}")
print(f"Desviación estándar de las recompensas: {std_reward}")

Promedio de recompensas: -0.387
Desviación estándar de las recompensas: 0.9017932135473187


1. El performance de la política aleatoria es bajo. El promedio de recompensas de -0.387 indica que, en promedio, la política aleatoria tiende a perder más juegos de los que gana.

2. Desviación Estándar (0.9018): Una desviación estándar relativamente alta muestra que los resultados de los episodios son variables, lo que es esperado con una política aleatoria, ya que las decisiones no siguen una estrategia coherente.

#### **2.1.3 Entrenamiento de modelo (0.2 puntos)**

A partir del siguiente [enlace](https://stable-baselines3.readthedocs.io/en/master/guide/algos.html), escoja un modelo de `stable_baselines3` y entrenelo para resolver el ambiente `Blackjack`.

In [ ]:
import gymnasium as gym
from gymnasium.spaces import MultiDiscrete
from stable_baselines3 import A2C
from stable_baselines3.common.vec_env import DummyVecEnv
import numpy as np

class FlattenObservation(gym.ObservationWrapper):
    def __init__(self, env):
        super(FlattenObservation, self).__init__(env)
        self.observation_space = MultiDiscrete([32, 11, 2])

    def observation(self, observation):
        return np.array(observation).flatten()

# Crear y envolver el ambiente
env = gym.make("Blackjack-v1")
env = FlattenObservation(env)
env = DummyVecEnv([lambda: env])

# Crear el modelo A2C
model = A2C("MlpPolicy", env, verbose=1, seed=42)

# Entrenar el modelo
model.learn(total_timesteps=50000)

# Guardar el modelo
model.save("a2c_blackjack")

# Cargar el modelo
model = A2C.load("a2c_blackjack")

Using cpu device
------------------------------------
| time/                 |          |
|    fps                | 212      |
|    iterations         | 100      |
|    time_elapsed       | 2        |
|    total_timesteps    | 500      |
| train/                |          |
|    entropy_loss       | -0.137   |
|    explained_variance | 0.24     |
|    learning_rate      | 0.0007   |
|    n_updates          | 99       |
|    policy_loss        | 0.00521  |
|    value_loss         | 0.737    |
------------------------------------
------------------------------------
| time/                 |          |
|    fps                | 278      |
|    iterations         | 200      |
|    time_elapsed       | 3        |
|    total_timesteps    | 1000     |
| train/                |          |
|    entropy_loss       | -0.144   |
|    explained_variance | 0.741    |
|    learning_rate      | 0.0007   |
|    n_updates          | 199      |
|    policy_loss        | -0.00609 |
|    value_loss      

#### **2.1.4 Evaluación de modelo (0.2 puntos)**

* Repita el ejercicio 2.1.2 pero utilizando el modelo entrenado.
* ¿Cómo es el performance de su agente?
* ¿Es mejor o peor que el escenario baseline?

In [ ]:
import gymnasium as gym
from gymnasium.spaces import MultiDiscrete
from stable_baselines3 import A2C
from stable_baselines3.common.vec_env import DummyVecEnv
import numpy as np

class FlattenObservation(gym.ObservationWrapper):
    def __init__(self, env):
        super(FlattenObservation, self).__init__(env)
        self.observation_space = MultiDiscrete([32, 11, 2])

    def observation(self, observation):
        return np.array(observation).flatten()

# Crear y envolver el ambiente
env = gym.make("Blackjack-v1")
env = FlattenObservation(env)
env = DummyVecEnv([lambda: env])

# Cargar el modelo entrenado
model = A2C.load("a2c_blackjack")

# Simular 5000 episodios con el modelo entrenado
n_episodes = 5000
rewards = []

for _ in range(n_episodes):
    done = False
    obs = env.reset()
    total_reward = 0
    while not done:
        action, _ = model.predict(obs, deterministic=True)
        obs, reward, done, _ = env.step(action)
        total_reward += reward[0]
    rewards.append(total_reward)

# Calcular promedio y desviación estándar
mean_reward = np.mean(rewards)
std_reward = np.std(rewards)

print(f"Promedio de recompensas: {mean_reward}")
print(f"Desviación estándar de las recompensas: {std_reward}")

Promedio de recompensas: -0.0712
Desviación estándar de las recompensas: 0.9466417273710259


Comparado con el escenario baseline, donde el promedio de recompensas era -0.387, el agente entrenado muestra una mejora, ya que la pérdida promedio es menor. Esto sugiere que el agente ha aprendido a tomar decisiones más efectivas que una política aleatoria.

#### **2.1.5 Estudio de acciones (0.2 puntos)**

* Genere una función que reciba un estado y retorne la accion del agente.
* Luego, use esta función para entregar la acción escogida frente a los siguientes escenarios:

  * Suma de cartas del agente es 6, dealer muestra un 7, agente no tiene tiene un as
  * Suma de cartas del agente es 19, dealer muestra un 3, agente tiene tiene un as

* ¿Son coherentes sus acciones con las reglas del juego?

Hint: ¿A que clase de python pertenecen los estados? Pruebe a usar el método `.reset` para saberlo.

In [ ]:
# escriba su respuesta acá
# Función para predecir la acción
def predict_action(state):
    state = np.array(state).flatten().reshape(1, -1)
    action, _ = model.predict(state, deterministic=True)
    return action

# Escenarios dados
state1 = (6, 7, 0)  # Suma de cartas del agente es 6, dealer muestra un 7, agente no tiene un as
state2 = (19, 3, 1)  # Suma de cartas del agente es 19, dealer muestra un 3, agente tiene un as

# Obtener las acciones
action1 = predict_action(state1)
action2 = predict_action(state2)

print(f"Acción para el estado 1: {action1}")
print(f"Acción para el estado 2: {action2}")

Acción para el estado 1: [1]
Acción para el estado 2: [0]


•	Estado 1: Acción [1] corresponde a “pedir carta”. Con una suma de 6 y el crupier mostrando un 7, es coherente que el agente pida carta, ya que la suma es muy baja.

•	Estado 2: Acción [0] corresponde a “plantarse”. Con una suma de 19 y el crupier mostrando un 3, es coherente plantarse, ya que la suma es alta y el riesgo de pasarse es considerable.

Ambas acciones son coherentes con las estrategias típicas en Blackjack.

### **2.2 LunarLander**

<p align="center">
  <img src="https://i.redd.it/097t6tk29zf51.jpg"
" width="400">
</p>

Similar a la sección 2.1, en esta sección usted se encargará de implementar una gente de RL que pueda resolver el ambiente `LunarLander`.


#### **2.2.1 Descripción de MDP (0.2 puntos)**


Comencemos preparando el ambiente:

In [ ]:
import gymnasium as gym
env = gym.make("LunarLander-v2", render_mode = "rgb_array", continuous = True) # notar el parámetro continuous = True

* Entregue una breve descripción sobre el ambiente [LunarLander](https://gymnasium.farama.org/environments/box2d/lunar_lander/) y su formulación en MDP, distinguiendo de forma clara y concisa los estados, acciones y recompensas.
* ¿Como se distinguen las acciones de este ambiente en comparación a `Blackjack`?
* En la preparación del ambiente se especifica el parámetro `continuous = True`. ¿Que implicancias tiene esto sobre el ambiente?

Descripción del ambiente Lunar Lander:

•	Estados: Es un vector de 8 dimensiones que incluye coordenadas, velocidades, ángulo, velocidad angular y contacto de las patas con el suelo.
•	Acciones: Cuatro acciones discretas (sin motor, motor principal, motor lateral izquierdo y derecho).
•	Recompensas: Basadas en la proximidad al objetivo, velocidad, inclinación, contacto de las patas y el uso de motores.

Comparación con Blackjack:

En Lunar Lander, las acciones son más complejas, involucrando control de motores y orientación, mientras que en Blackjack solo son decisiones de “pedir” o “plantarse”.

Implicancias de continuous = True:

Cuando continuous = True, las acciones se vuelven continuas, permitiendo un control más preciso del motor principal y los motores laterales, en lugar de simplemente encender o apagar motores.

#### **2.2.2 Generando un Baseline (0.2 puntos)**

* Simule un escenario en donde se escojan acciones aleatorias. Repita esta simulación 10 veces y reporte el promedio y desviación de las recompensas.
* ¿Cómo calificaría el performance de esta política?

In [ ]:
import gymnasium as gym
import numpy as np

# Crear el ambiente
env = gym.make("LunarLander-v2", render_mode="rgb_array", continuous=True)

# Simular 10 episodios con acciones aleatorias
n_episodes = 10
rewards = []

for _ in range(n_episodes):
    done = False
    obs, _ = env.reset()
    total_reward = 0
    while not done:
        action = env.action_space.sample()  # Acción aleatoria
        obs, reward, done, _, _ = env.step(action)
        total_reward += reward
    rewards.append(total_reward)

# Calcular promedio y desviación estándar
mean_reward = np.mean(rewards)
std_reward = np.std(rewards)

print(f"Promedio de recompensas: {mean_reward}")
print(f"Desviación estándar de las recompensas: {std_reward}")

Promedio de recompensas: -246.5308163586496
Desviación estándar de las recompensas: 211.09166153352746


El performance de la política aleatoria es bajo, con un promedio de recompensas de aproximadamente -246.53. Esto indica que, en general, el agente tiene dificultades para aterrizar correctamente. La alta desviación estándar de 211.09 sugiere una gran variabilidad en los resultados, lo que es típico de una política que no toma decisiones informadas.

#### **2.2.3 Entrenamiento de modelo (0.2 puntos)**

* A partir del siguiente [enlace](https://stable-baselines3.readthedocs.io/en/master/guide/algos.html), escoja un modelo de `stable_baselines3` y entrenelo para resolver el ambiente `LunarLander` **usando 10000 timesteps de entrenamiento**.

In [ ]:
import gymnasium as gym
from stable_baselines3 import DDPG
from stable_baselines3.common.vec_env import DummyVecEnv

# Crear y envolver el ambiente
env = gym.make("LunarLander-v2", continuous=True)
env = DummyVecEnv([lambda: env])

# Crear el modelo DDPG
model = DDPG("MlpPolicy", env, verbose=1, seed=42)

# Entrenar el modelo
model.learn(total_timesteps=10000)

# Guardar el modelo
model.save("ddpg_lunar_lander")

# Cargar el modelo
model = DDPG.load("ddpg_lunar_lander")



Using cpu device
---------------------------------
| time/              |          |
|    episodes        | 4        |
|    fps             | 51       |
|    time_elapsed    | 7        |
|    total_timesteps | 381      |
| train/             |          |
|    actor_loss      | 10.7     |
|    critic_loss     | 21.5     |
|    learning_rate   | 0.001    |
|    n_updates       | 280      |
---------------------------------
---------------------------------
| time/              |          |
|    episodes        | 8        |
|    fps             | 38       |
|    time_elapsed    | 17       |
|    total_timesteps | 683      |
| train/             |          |
|    actor_loss      | 23.4     |
|    critic_loss     | 22.4     |
|    learning_rate   | 0.001    |
|    n_updates       | 582      |
---------------------------------
---------------------------------
| time/              |          |
|    episodes        | 12       |
|    fps             | 31       |
|    time_elapsed    | 62      

#### **2.2.4 Evaluación de modelo (0.2 puntos)**

* Repita el ejercicio 2.2.2 pero utilizando el modelo entrenado.
* ¿Cómo es el performance de su agente? ¿Es mejor o peor que el escenario baseline?

In [ ]:
import gymnasium as gym
import numpy as np
from stable_baselines3 import DDPG
from stable_baselines3.common.vec_env import DummyVecEnv

# Crear y envolver el ambiente
env = gym.make("LunarLander-v2", continuous=True)
env = DummyVecEnv([lambda: env])

# Cargar el modelo entrenado
model = DDPG.load("ddpg_lunar_lander")

# Simular 10 episodios con el modelo entrenado
n_episodes = 10
rewards = []

for _ in range(n_episodes):
    done = False
    obs = env.reset()
    total_reward = 0
    while not done:
        action, _ = model.predict(obs, deterministic=True)
        obs, reward, done, _ = env.step(action)
        total_reward += reward[0]
    rewards.append(total_reward)

# Calcular promedio y desviación estándar
mean_reward = np.mean(rewards)
std_reward = np.std(rewards)

print(f"Promedio de recompensas: {mean_reward}")
print(f"Desviación estándar de las recompensas: {std_reward}")

Promedio de recompensas: -76.05608874436003
Desviación estándar de las recompensas: 48.65351366634402


Performance del Agente:
El agente entrenado tiene un promedio de recompensas de -76.06 con una desviación estándar de 48.65. Esto muestra una mejora significativa respecto al baseline.

Comparación con el Escenario Baseline:
Comparado con el baseline, donde el promedio de recompensas era -246.53, el agente es claramente mejor. Las recompensas más altas y la menor desviación estándar indican un comportamiento más consistente y efectivo.

#### **2.2.5 Optimización de modelo (0.2 puntos)**

* Repita los ejercicios 2.2.3 y 2.2.4 hasta obtener un nivel de recompensas promedio mayor a 50. Para esto, puede cambiar manualmente parámetros como:
  - `total_timesteps`
  - `learning_rate`
  - `batch_size`

* Una vez optimizado el modelo, use la función `export_gif` entregada para estudiar el comportamiento de su agente en la resolución del ambiente, comente sobre sus resultados.

* Adjunte el gif generado en su entrega. Si, además, adjuntan el gif en el markdown tendrán un bonus de 0.1.

In [ ]:
import imageio
import numpy as np

def export_gif(model, env, n=5):
    '''
    Función que exporta a gif el comportamiento del agente en n episodios.
    '''
    images = []
    for episode in range(n):
        obs = env.reset()
        done = False
        while not done:
            img = env.render(mode="rgb_array")
            if img is not None:
                images.append(img)
            action, _ = model.predict(obs, deterministic=True)
            obs, reward, done, _ = env.step(action)

    if images:
        imageio.mimsave("agent_performance.gif", [np.array(img) for i, img in enumerate(images) if i % 2 == 0], fps=29)


In [ ]:
#Creando un modelo que cambie los parametros.
import gymnasium as gym
from stable_baselines3 import DDPG
from stable_baselines3.common.vec_env import DummyVecEnv

# Crear y envolver el ambiente
env = gym.make("LunarLander-v2", continuous=True)
env = DummyVecEnv([lambda: env])

# Ajustar parámetros del modelo
model = DDPG(
    "MlpPolicy",
    env,
    verbose=1,
    seed=42,
    learning_rate=0.0001,  # Ajusta el learning rate
    batch_size=64         # Ajusta el batch size
)

# Entrenar el modelo
model.learn(total_timesteps=20000)  # Aumenta el total_timesteps

# Guardar el modelo
model.save("ddpg_lunar_lander_optimized")

# Cargar el modelo
model = DDPG.load("ddpg_lunar_lander_optimized")

Using cpu device
---------------------------------
| time/              |          |
|    episodes        | 4        |
|    fps             | 73       |
|    time_elapsed    | 4        |
|    total_timesteps | 354      |
| train/             |          |
|    actor_loss      | 5.97     |
|    critic_loss     | 105      |
|    learning_rate   | 0.0001   |
|    n_updates       | 253      |
---------------------------------
---------------------------------
| time/              |          |
|    episodes        | 8        |
|    fps             | 70       |
|    time_elapsed    | 9        |
|    total_timesteps | 669      |
| train/             |          |
|    actor_loss      | 6.45     |
|    critic_loss     | 124      |
|    learning_rate   | 0.0001   |
|    n_updates       | 568      |
---------------------------------
---------------------------------
| time/              |          |
|    episodes        | 12       |
|    fps             | 63       |
|    time_elapsed    | 19      

In [ ]:
import gymnasium as gym
import numpy as np
from stable_baselines3 import DDPG
from stable_baselines3.common.vec_env import DummyVecEnv

# Crear y envolver el ambiente
env = gym.make("LunarLander-v2", continuous=True)
env = DummyVecEnv([lambda: env])

# Cargar el modelo
model = DDPG.load("ddpg_lunar_lander_optimized")

# Simular 10 episodios con el modelo entrenado
n_episodes = 10
rewards = []

for _ in range(n_episodes):
    done = False
    obs = env.reset()
    total_reward = 0
    while not done:
        action, _ = model.predict(obs, deterministic=True)
        obs, reward, done, _ = env.step(action)
        total_reward += reward[0]
    rewards.append(total_reward)

# Calcular promedio y desviación estándar
mean_reward = np.mean(rewards)
std_reward = np.std(rewards)

print(f"Promedio de recompensas: {mean_reward}")
print(f"Desviación estándar de las recompensas: {std_reward}")

Promedio de recompensas: -58.281686818308664
Desviación estándar de las recompensas: 64.18256240944172


In [ ]:

#usando la función
export_gif(model, env, n=5)